Userland/Libraries/LibDSP/Resampler.h

```diff
@@ -0,0 +1,198 @@
+        double const y2 = m_sinc_table[n + 1];
+
+        return y1 + fraction * (y2 - y1);
+    }
+
+private:
+    Vector<double> m_sinc_table;
+    size_t const m_sinc_taps;
+    size_t const m_sinc_lookup_oversample;
+};
+
+template<typename TSinc>
+class SincResampler {
+public:
+    using SampleType = float;
+
+    SincResampler(u32 rate_from,
+        u32 rate_to,
+        size_t max_input_buffer_size,
+        size_t sinc_taps = 101,
+        size_t sinc_lookup_oversample = 1000)
+        : m_sinc_function(sinc_taps, sinc_lookup_oversample)
+    {
+        m_phase = 0;
+        m_sinc_taps = sinc_taps;
+        m_processed_sample_count = 0;
+        m_ratio = rate_from / static_cast<double>(rate_to);
+
+        double const transition_bandwidth = 2000 / static_cast<double>(rate_from);
+        double const cutoff = 1.0 / (2.0 * m_ratio);
+        MUST(calculate_lowpass(cutoff, transition_bandwidth));
+
+        // The input buffer is a shifting window over all the input data.
+        // When we shift the window to the right, we have to leave room for the previous
+        // `sinc_taps` number of samples to be multiplied with the left side of the sinc.
+        m_input_buffer_size = max_input_buffer_size + 2 * sinc_taps + 1;
+        m_input_buffer.resize(m_input_buffer_size);
+    }
+
+    size_t process(Span<SampleType> input, Span<SampleType> output)
+    {
+        size_t const total_tap_count = 2 * m_sinc_taps + 1;
+        Span<SampleType> old_data { &m_input_buffer[m_input_buffer.size() - total_tap_count], total_tap_count };
+        Span<SampleType> new_data { &m_input_buffer[total_tap_count], input.size() };
+
+        memmove(m_input_buffer.data(), old_data.data(), old_data.size() * sizeof(SampleType));
+        memmove(new_data.data(), input.data(), input.size() * sizeof(SampleType));
+        m_input_buffer_size = old_data.size() + input.size();
+
+        assert(m_input_buffer_size <= m_input_buffer.size());
+
+        if (m_ratio > 1) {
+            for (size_t i = 0; i < input.size(); i++) {
+                new_data[i] = static_cast<SampleType>(m_lowpass->process(new_data[i]));
+            }
+        }
+
+        size_t samples_written = 0;
+        size_t const input_buffer_start_sample_count = static_cast<size_t>(m_processed_sample_count * m_ratio);
+
+        while (true) {
+            size_t const n = m_sinc_taps + static_cast<size_t>(m_processed_sample_count * m_ratio) - input_buffer_start_sample_count;
+
+            // +1 because the break condition for the for loop is `<=`, not `<`.
+            if (n + m_sinc_taps + 1 >= m_input_buffer_size) {
+                break;
+            }
+
+            // Whittaker–Shannon interpolation formula
+            SampleType sum = 0;
+            for (i32 k = -m_sinc_taps; k <= static_cast<i32>(m_sinc_taps); k++) {
+                sum += static_cast<SampleType>(static_cast<double>(m_input_buffer[n + k]) * m_sinc_function(m_phase - k));
```

Comment:
Then I get a compiler error about an implicit conversion of one of the operands. Either way I'll try to use the `Sample` type, maybe the issue will just disappear until someone else needs to use a different type for their samples or blocks or whatever.

Reply:
> Then I get a compiler error about an implicit conversion of one of the operands. Either way I'll try to use the `Sample` type, maybe the issue will just disappear until someone else needs to use a different type for their samples or blocks or whatever.

I think you'd have to cast the sinc function to `float` instead, since Sample mainly interoperates with `float` for memory efficiency reasons. (It's fine to use double in intermediate steps, but we need to cast back to float in most applications)